### PR TITLE
Fix/issue 5351 2

### PR DESCRIPTION
**Pull request by rajyan**
fixes https://github.com/phpstan/phpstan/issues/4885

Issue was not fixed by https://github.com/phpstan/phpstan-src/pull/1063, because`ensureShallowNonNullability` is not handling optional keys. This PR specifies type for `ArrayDimFetch` to make the `nonNullable` scope more precise.

### CODE REVIEW

**Comment by rajyan** [comment]
Maybe it is better to solve this one first
https://github.com/phpstan/phpstan/issues/6870
and replace the `Coalesce` to simulate `isset($leftExpr) ? $leftExpr : $rightExpr`

File: tests/PHPStan/Analyser/data/bug-4885.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?php // lint >= 8.0
+
+namespace Bug4885;
+
+use function PHPStan\Testing\assertType;
+
+class Foo
+{
+	/** @param array{word?: string} $data */
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/bug-4885.php:9]: Please also test `array{word?: string|null}`.

File: src/Analyser/NodeScopeResolver.php [hunk 2]
```diff
@@ -1561,23 +1561,35 @@ private function ensureShallowNonNullability(MutatingScope $scope, Expr $exprToS
...
 
-			return new EnsuredNonNullabilityResult(
-				$scope,
-				[
-					new EnsuredNonNullabilityResultExpression($exprToSpecify, $exprType, $nativeType),
-				],
-			);
+		if ($exprToSpecify instanceof ArrayDimFetch && $exprToSpecify->dim !== null) {
+			$dimFetchVarType = $scope->getType($exprToSpecify->var);
+			if ($dimFetchVarType instanceof ConstantArrayType) {
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php:1577]: I worry this logic isn't really sufficient. You can also have a union of constant arrays. And I guess it should also do something aboout nested ArrayDimFetch - `$a['a']['b']['c']` where the intermediate keys are also optional...

I feel like instead of reimplementing the logic here, this place should call TypeSpecifier because the place where it handles `Isset_` looks like it has this covered: https://github.com/phpstan/phpstan-src/blob/5c48a180115306c5ca005c2617d8b27db1c2df49/src/Analyser/TypeSpecifier.php#L742-L826

So the scope after `$x ?? throw` should be something like `$scope->filterByTruthyValue(new Isset_($expr->left))` (which delegates the call to TypeSpecifier internally).

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by rajyan** [comment]
@ondrejmirtes 
Thank you for the review.
I was having the same feeling and thought maybe the whole `Coalesce` process can be replaced with `isset($leftExpr) ? $leftExpr : $rightExpr`. With this, the scope should work perfect with `NeverType` and can filter by `TruthyScope` without any additional implementation (Also, https://github.com/phpstan/phpstan-src/pull/1063 can be reverted I think). Although, early termination for non `Stmt` is not working now, https://github.com/phpstan/phpstan/issues/6870
so maybe I would fix that first.

**Comment by ondrejmirtes** [comment]
Try whatever you think should work 😊

**Comment by rajyan** [comment]
move to https://github.com/phpstan/phpstan-src/pull/1104
